src/npe2/io_utils.py

```diff
@@ -154,9 +154,13 @@ def _read(
     if _pm is None:
         _pm = PluginManager.instance()
 
-    for rdr in _pm.iter_compatible_readers(paths):
-        if plugin_name and rdr.plugin_name != plugin_name:
-            continue
+    # get readers compatible with paths and chosen plugin - raise errors if
+    # choices are invalid or there's nothing to try
+    chosen_compatible_readers = _get_compatible_readers_by_choice(
+        plugin_name, paths, _pm
+    )
```

Comment:
The comment is helpful. Consider adding an assert after line 161 that sort of checks the invariant holds - it's the kind of thing that would save me some time debugging later on if I broke the rule.

```suggestion
    # get readers compatible with paths and chosen plugin - raise errors if
    # choices are invalid or there's nothing to try
    chosen_compatible_readers = _get_compatible_readers_by_choice(
        plugin_name, paths, _pm
    )
    assert chosen_compatible_readers, "Expected an exception before this point"
```